### PR TITLE
Add ABI stress tests

### PR DESCRIPTION
**Pull request by jakobbotsch**
This adds stress tests that dynamically generate methods with different
signatures and tests calls to them. It tests both tailcalls to them and
pinvokes to them (caller -> pinvoke -> reverse pinvoke -> callee).

This adds 2 tests: one for pinvokes (testing caller -> pinvoke -> reverse pinvoke -> callee) and one for tailcalls. Each of these run different debug/optimizations modes, so in total this is 8 tests. I have aimed for these to take roughly 10 seconds each (x64 checked on an Ubuntu VM), so this adds around 80 secs of CPU time to the pri0 tests. I am curious if anyone has thoughts on whether this is an acceptable amount of time for these tests or whether the total number of calls should be decreased. Right now each test is testing 1000 calls.

cc @dotnet/jit-contrib

### CODE REVIEW

**Comment by jakobbotsch** [comment]
ARM32 needs some work. The x86 failure seems like a legitimate bug:
```
Assert failure(PID 904 [0x00000388], Thread: 1584 [0x0630]): Assertion failed '!"Unexpected node in impNormStructVal()"' in 'ILStubClass:IL_STUB_PInvoke(struct,double,struct,struct,struct,double,struct,struct,int,double,int,struct,double,struct,short,struct,struct,struct,double):int' (IL size 210)

    File: f:\workspace.9\_work\1\s\src\jit\importer.cpp Line: 1772
    Image: C:\dotnetbuild\work\0ea89f38-d444-41ae-95ab-90bd1ab45f4b\Payload\CoreRun.exe
```
and also
```
Assert failure(PID 3236 [0x00000ca4], Thread: 3536 [0x0dd0]): Assertion failed 'thisPtr->TypeGet() == TYP_I_IMPL || thisPtr->TypeGet() == TYP_BYREF' in 'DynamicClass:ABIStress_PInvoker0(ubyte,struct,float,short,int,struct,struct):ref' (IL size 1555)

    File: f:\workspace.9\_work\1\s\src\jit\importer.cpp Line: 6639
    Image: C:\dotnetbuild\work\0ea89f38-d444-41ae-95ab-90bd1ab45f4b\Payload\CoreRun.exe
```

I will take a closer look tomorrow.

**Comment by jashook** [comment]
Opened #26101

File: tests/src/JIT/Stress/ABI/ABIs.cs [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+using System.Numerics;
+using System.Runtime.InteropServices;
+using System.Runtime.Intrinsics;
+using System.Text;
+
+namespace ABIStress
+{
+    internal interface IAbi
+    {
+        Type[] TailCalleeCandidateArgTypes { get; }
```
Comment (jashook) [inline tests/src/JIT/Stress/ABI/ABIs.cs]: Please add some comments describing the purpose for these interface methods.

File: tests/src/JIT/Stress/ABI/ABIs.cs [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+                typeof(S31U), typeof(S32U),
+            };
+
+        public CallingConvention[] PInvokeConventions { get; } = { CallingConvention.Cdecl, CallingConvention.StdCall, CallingConvention.ThisCall, };
+
+        public int ApproximateArgStackAreaSize(List<TypeEx> parameters)
+        {
+            int size = 0;
+            foreach (TypeEx pm in parameters)
+                size += (pm.Size + 3) & ~3;
```
Comment (jashook) [inline tests/src/JIT/Stress/ABI/ABIs.cs]: Can you add a roundup method and explain what this is accomplishing. Was hard to follow just reading through the code.

File: tests/src/JIT/Stress/ABI/ABIs.cs [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+        public CallingConvention[] PInvokeConventions { get; } = { CallingConvention.Cdecl };
+
+        public int ApproximateArgStackAreaSize(List<TypeEx> parameters)
+        {
+            int size = 0;
+            foreach (TypeEx pm in parameters)
+            {
+                // 1, 2, 4 and 8 byte structs are passed directly by value, everything else
+                // by ref.
```
Comment (jashook) [inline tests/src/JIT/Stress/ABI/ABIs.cs]: nit: `by ref` -> `byref`

File: tests/src/JIT/Stress/ABI/ABIs.cs [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+            foreach (TypeEx pm in parameters)
+                size += (pm.Size + 7) & ~7;
+
+            return size;
+        }
+    }
+
+    internal class Arm64Abi : IAbi
+    {
+        // For Arm64 everything can be passed everything by value.
```
Comment (jashook) [inline tests/src/JIT/Stress/ABI/ABIs.cs]: `>16 byte` structs I believe are passed byref

File: tests/src/JIT/Stress/ABI/Gen.cs [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+                return GenConstantVector<Vector128<int>, int>(rand);
+
+            if (type == typeof(Vector256<int>))
+                return GenConstantVector<Vector256<int>, int>(rand);
+
+            Debug.Assert(fields != null);
+            return Activator.CreateInstance(type, fields.Select(fi => GenConstant(fi.FieldType, null, rand)).ToArray());
+        }
+    }
+    internal abstract class Value
```
Comment (jashook) [inline tests/src/JIT/Stress/ABI/Gen.cs]: super nit: missing new line

File: tests/src/JIT/Stress/ABI/Gen.cs [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+using System.Linq;
+using System.Numerics;
+using System.Reflection;
+using System.Reflection.Emit;
+using System.Runtime.CompilerServices;
+using System.Runtime.Intrinsics;
+
+namespace ABIStress
+{
+    internal static class Gen
```
Comment (jashook) [inline tests/src/JIT/Stress/ABI/Gen.cs]: I think it is worth adding a comment, explaining what the high level overview of the class is doing. Either that or a small blurb on each member method.

**Comment by jashook** [comment]
Please mark these tests as <JitOptimizationSensitive>true</JitOptimizationSensitive> and <GCStressIncompatible>true</GCStressIncompatible>

File: tests/src/JIT/Stress/ABI/Program.cs [hunk 7]
```diff
@@ -0,0 +1,653 @@
...
+                resultIndex++;
+            }
+
+            g.Emit(OpCodes.Ldloc, resultsArrLocal);
+            g.Emit(OpCodes.Ret);
+
+            (object callerResult, object calleeResult) =
+                InvokeCallerCallee(caller, pms, callee.Method, args, rand);
+
+            GC.KeepAlive(delegates);
```
Comment (jashook) [inline tests/src/JIT/Stress/ABI/Program.cs]: Please add a comment explaining why this is necessary. Also it probably warrants a TODO remove when #... is fixed.

Comment (jakobbotsch) [reply tests/src/JIT/Stress/ABI/Program.cs]: Delegates for which you have called `Marshal.GetFunctionPointerForDelegate` need to be kept rooted since the native pointer will point to garbage once the delegate has been collected (this is not a bug). I will add a comment.

Comment (jashook) [reply tests/src/JIT/Stress/ABI/Program.cs]: Gotcha thanks!

**Comment by jakobbotsch** [comment]
I have some local comment changes. I will hold off with pushing them to see if there is other feedback first.

File: tests/src/JIT/Stress/ABI/TailCallEventListener.cs [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        public int NumCallersSeen { get; set; }
+        public int NumSuccessfulTailCalls { get; set; }
+        public Dictionary<string, int> FailureReasons { get; } = new Dictionary<string, int>();
+
+        protected override void OnEventSourceCreated(EventSource eventSource)
+        {
+            if (eventSource.Name != "Microsoft-Windows-DotNETRuntime")
+                return;
+
+            EventKeywords jitTracing = (EventKeywords)0x1000; // JITSymbols | JITTracing
```
[the remark this answers is not in the document]
Comment (jakobbotsch) [reply tests/src/JIT/Stress/ABI/TailCallEventListener.cs]: I found them in this document:
https://github.com/dotnet/diagnostics/blob/master/documentation/dotnet-trace-instructions.md#commonly-used-keywords-for-the-microsoft-windows-dotnetruntime-provider

Note that the comment is outdated. I will fix the comment.

**Comment by jashook** [comment]
Ci has random failure. I believe this change is ready to go. Ping @dotnet/jit-contrib

**Comment by jashook** [comment]
As this change is hoped to validate https://github.com/dotnet/coreclr/pull/25932 and it has two approvals. I am inclined to just merge. Feel free to make suggestions after the merge, things can be changed in the future.

**Comment by BruceForstall** [comment]
@jakobbotsch The `_r` and `_ro` proj files are the same. The `_ro` version needs `<Optimize>True</Optimize>`

**Comment by jakobbotsch** [comment]
Well spotted, opened #26183. Thanks!

**Comment by BruceForstall** [comment]
Actually, I have a question: do we need all 4 versions for each? It seems like the key testing here is what the JIT does for the generated code, not whether the test generator itself is debug/release or debuggable/not-debuggable. Maybe just the equivalent of `_ro` is all that is needed?

**Comment by jakobbotsch** [comment]
I must admit I do not totally understand how the different options affect the JIT's codegen. The test places the generated methods in the module of the test's class:
https://github.com/dotnet/coreclr/blob/bc1bf27bd1b58a34b169b77988b74e58273be2a4/tests/src/JIT/Stress/ABI/Program.cs#L220-L221
I am unsure if this will make the JIT treats these dynamic methods like it would the test's methods, however. Do you have pointers for how I would check?

**Comment by BruceForstall** [comment]
Maybe you could generate a single test from each test flavor, and look at the JitDump for each. The header of the JitDump should show what kind of code is being generated (e.g., MinOpts versus full opts). You could also see if the generated JIT asm differs.

**Comment by jakobbotsch** [comment]
These are the results. I have captured both the `Program.Main` method (part of the test) and the `ABIStress_PInvoker0` method (generated by the test) when running the wrapper scripts generated for the tests:
```
------------------- _d version ------------------
****** START compiling ABIStress.Program:Main(ref):int (MethodHash=3f1384e6)
Generating code for Unix x64
OPTIONS: compCodeOpt = BLENDED_CODE
OPTIONS: compDbgCode = true
OPTIONS: compDbgInfo = true
OPTIONS: compDbgEnC  = false
OPTIONS: compProcedureSplitting   = false
OPTIONS: compProcedureSplittingEH = false

****** START compiling DynamicClass:ABIStress_PInvoker0(ubyte,struct,float,short,struct,struct,struct):ref (MethodHash=82c6f3b1)
Generating code for Unix x64
OPTIONS: compCodeOpt = BLENDED_CODE
OPTIONS: compDbgCode = true
OPTIONS: compDbgInfo = true
OPTIONS: compDbgEnC  = false
OPTIONS: compProcedureSplitting   = false
OPTIONS: compProcedureSplittingEH = false


------------------- _do version ------------------
****** START compiling ABIStress.Program:Main(ref):int (MethodHash=6c172069)
Generating code for Unix x64
OPTIONS: compCodeOpt = BLENDED_CODE
OPTIONS: compDbgCode = false
OPTIONS: compDbgInfo = true
OPTIONS: compDbgEnC  = false
OPTIONS: compProcedureSplitting   = false
OPTIONS: compProcedureSplittingEH = false
****** START compiling DynamicClass:ABIStress_PInvoker0(ubyte,struct,float,short,struct,struct,struct):ref (MethodHash=c1da3dfe)
Generating code for Unix x64
OPTIONS: compCodeOpt = BLENDED_CODE
OPTIONS: compDbgCode = false
OPTIONS: compDbgInfo = true
OPTIONS: compDbgEnC  = false
OPTIONS: compProcedureSplitting   = false
OPTIONS: compProcedureSplittingEH = false


------------------- _r version ------------------
****** START compiling ABIStress.Program:Main(ref):int (MethodHash=3f13d0f0)
Generating code for Unix x64
OPTIONS: compCodeOpt = BLENDED_CODE
OPTIONS: compDbgCode = true
OPTIONS: compDbgInfo = true
OPTIONS: compDbgEnC  = false
OPTIONS: compProcedureSplitting   = false
OPTIONS: compProcedureSplittingEH = false

****** START compiling DynamicClass:ABIStress_PInvoker0(ubyte,struct,float,short,struct,struct,struct):ref (MethodHash=82c6c5a7)
Generating code for Unix x64
OPTIONS: compCodeOpt = BLENDED_CODE
OPTIONS: compDbgCode = true
OPTIONS: compDbgInfo = true
OPTIONS: compDbgEnC  = false
OPTIONS: compProcedureSplitting   = false
OPTIONS: compProcedureSplittingEH = false


------------------- _ro version ------------------
****** START compiling ABIStress.Program:Main(ref):int (MethodHash=6c0f6b3f)
Generating code for Unix x64
OPTIONS: compCodeOpt = BLENDED_CODE
OPTIONS: compDbgCode = false
OPTIONS: compDbgInfo = true
OPTIONS: compDbgEnC  = false
OPTIONS: compProcedureSplitting   = false
OPTIONS: compProcedureSplittingEH = false
****** START compiling DynamicClass:ABIStress_PInvoker0(ubyte,struct,float,short,struct,struct,struct):ref (MethodHash=c1dfe2a8)
Generating code for Unix x64
OPTIONS: compCodeOpt = BLENDED_CODE
OPTIONS: compDbgCode = false
OPTIONS: compDbgInfo = true
OPTIONS: compDbgEnC  = false
OPTIONS: compProcedureSplitting   = false
OPTIONS: compProcedureSplittingEH = false
```

So it looks like the dynamic methods inherit the options from the main module. On the other hand I don't see anything about `minOpts` in any of the methods? `_d` = `_r` and `_do` = `_ro` in terms of the options header, even for the main method directly in the test. Not exactly what I expected.

**Comment by BruceForstall** [comment]
So it looks like the `<DebugType>` doesn't have any interesting effect (at least here). I can't remember now what the effect is.

MinOpts might not be specifically set, but we will generate unoptimized code for `compDbgCode = true`:
```
 bool OptimizationDisabled()
        {
            return MinOpts() || compDbgCode;
        }
```

**Comment by jakobbotsch** [comment]
It looks like there might be some test cleanup to be done:
https://github.com/dotnet/corefx/issues/28301#issuecomment-374996117

**Comment by jakobbotsch** [comment]
Some tests use `<DebugType>None</DebugType>`, but from what I can tell this does not affect the JIT's codegen either (`compDbgInfo` still reports true). And according to Roslyn `PdbOnly` and `Full` is the same thing: https://github.com/dotnet/roslyn/blob/master/docs/compilers/CSharp/CommandLine.md

I can clean it up. What do you think about the following:
1. Merge all `_d` and `_r` into a `_d` that has
```xml
<DebugType>Full</DebugType>
<Optimize>False</Optimize>
```
2. Merge all `_do` and `_ro` versions into a single `_r` that has
```xml
<DebugType>Full</DebugType>
<Optimize>True</Optimize>
```

**Comment by BruceForstall** [comment]
Wouldn't just removing the `_r` and `_ro` have the same effect? (i.e., leave `_d` and `_do`)

**Comment by jakobbotsch** [comment]
Yes it would, although the naming is not super obvious to me if the debuggability does not affect anything in the JIT. But I am fine with that -- should I go ahead and do so?

**Comment by CarolEidt** [comment]
I have also found the names less than mnemonic. I don't have a strong opinion, but to me  something like _o and _no would make more sense, since both specify the same `<DebugType>`.

**Comment by BruceForstall** [comment]
The names have confused me in the past, also. But they're also pervasive in the tree, currently. So yes, I'd suggest just getting rid of the `_r` and `_ro` versions for this test.

**Comment by CarolEidt** [comment]
Sounds reasonable

**Comment by jakobbotsch** [comment]
I was planning on doing some more widespread cleanup, otherwise my tests will just be inconsistent with the rest of the JIT tree. I.e. I was hoping to clean up all the tests in the JIT test tree to follow consistent conventions (_o and _no would be reasonable to me too).

**Comment by BruceForstall** [comment]
While that's admirable, it's my impression that your tests already follow the existing conventions. And I don't think it's worth your time to worry about "fixing" the rest of the tests to conform to some new convention.

**Comment by jakobbotsch** [comment]
From some offline talk it seems we do have significant overhead for no gain with having all these 4 variants so I will try to fix this up. My suggestion for the naming scheme is `_opt` and `_unopt` to avoid any confusion and make it easier to spot in logs. What do you think about that convention @CarolEidt?

**Comment by CarolEidt** [comment]
IMO `_opt` and `_unopt` would be an improvement. I guess the question is whether the impact of simply deleting the `_r` and `_ro` variants would be significantly less.

**Comment by AndyAyersMS** [comment]
I could be wrong, but my recollection is that there really are 4 (and perhaps even 5) different IL patterns C# can emit for a method. So it would be good to run some IL diffs across the current variants before we decide some of them are redundant.

@VSadov I remember we talked about this at one point long ago -- do you recall?

**Comment by VSadov** [comment]
Is this about optimizations?

**Comment by VSadov** [comment]
Optimizations in C# mostly affect locals and control flow. Also has effect on lowering of lambdas/async,etc..

If this is about parameters, there is not a lot of flexibility there.

**Comment by jakobbotsch** [comment]
We are curious about whether `DebugType` affects the code generated by Roslyn or whether it purely affects the generation of PDBs (and if we can rely on this).

**Comment by VSadov** [comment]
On the side of locals - I think there is basically 3 modes. Unless something changed.

optimize +/-   
these will will differ significantly. 

lowering may introduce a lot of temps and most will stay when not optimizing. 
/o+ will try to eliminate locals where possible.

then there is a combination of   "/o+ /debug+"  -   it is basically same as /o+ , but will suppress some locals-elimination rules. I do not remember all the details exactly. I think it forces IL stack to be empty between statements. 
It was done for live debugging where breakpoints need to have empty stack, but otherwise want to work with optimized code.

**Comment by jakobbotsch** [comment]
I think that means our tests should stay as they are if the debug type potentially affects the IL generated. Thanks for the response @VSadov!

**Comment by VSadov** [comment]
confusingly and for some historical reasons 
`-debug[+|-]  `         and    ` -debug:{full|pdbonly|portable|embedded} `      are  different settings.

The `[+|-]` will have effect on emit. The other is just the format of the debug info,

**Comment by jakobbotsch** [comment]
I see. Then it will take some tracking down to see how our `DebugType` affects it -- i.e. whether the `-debug+` or `-debug-` flags are ever emitted. I wonder if we have a hole here...

**Comment by VSadov** [comment]
Also note that cases where debug [+|-] will have effect on emit are fairly rare. Do not be surprised when nothing changes.

**Comment by jakobbotsch** [comment]
Here is the difference between `<DebugType>None</DebugType>` and `<DebugType>Full</DebugType>` for our invocation of csc:
```diff
@@ -1,4 +1,4 @@
-<DebugType>Full</DebugType>
+<DebugType>None</DebugType>
 /ssd/jakob/coreclr/.dotnet/dotnet
 exec
 "/ssd/jakob/coreclr/.packages/microsoft.net.compilers.toolset/3.2.0-beta2-19272-03/build/../tasks/netcoreapp2.1/bincore/csc.dll"
@@ -11,11 +11,10 @@ exec
 /errorreport:prompt
 /warn:4
 /define:DEBUG;TRACE;XUNIT_PERF;TRACE;FEATURE_ARRAYSTUB_AS_IL;FEATURE_MULTICASTSTUB_AS_IL;FEATURE_STUBS_AS_IL;FEATURE_COLLECTIBLE_ALC;FEATURE_MANAGED_ETW;FEATURE_MANAGED_ETW_CHANNELS;FEATURE_PAL;FEATURE_PERFTRACING;FEATURE_EVENTSOURCE_XPLAT;FEATURE_DEFAULT_INTERFACES;FEATURE_TYPEEQUIVALENCE;FEATURE_BASICFREEZE;PROFILING_SUPPORTED;PLATFORM_UNIX;CHECKED;NETCOREAPP;NETCOREAPP3_0
-/debug+
-/debug:Full
+/debug-
 /filealign:512
 /optimize-
-/out:/ssd/jakob/coreclr/tests/../bin/tests/obj/Linux.x64.Checked/Managed/JIT/Stress/ABI/pinvokes_d/pinvokes_d.dll
+/out:/ssd/jakob/coreclr/tests/../bin/tests/obj/Linux.x64.Checked/Managed/JIT/Stress/ABI/pinvokes_r/pinvokes_r.dll
 /target:exe
 /warnaserror-
 /utf8output
```

So, it would seem it is beneficial to have all 4 variants for tests written in C#. In my particular case (ABI stress tests) I should only have _d and _do versions because I generate the IL. Thanks for the pointers @VSadov.
